### PR TITLE
Market resource prices

### PR DESCRIPTION
**Pull request by MaanooAk**
This market resource buy and sell logic (not the building)

- The global base prices
- Calc prices for selling and buying
- Perform buy and sell

Based on [doc/reverse_engineering/market.md#prices](https://github.com/SFTtech/openage/blob/master/doc/reverse_engineering/market.md#prices).

### CODE REVIEW

**Comment by TheJJ** [comment]
Can you add documentation about the price calculation please?

**Comment by MaanooAk** [comment]
Its on the [doc/reverse_engineering/market.md#prices](https://github.com/SFTtech/openage/blob/master/doc/reverse_engineering/market.md#prices) and its very simple:
```
buy -> increase value
sell -> decrease value
```
and there is a min and max value. Are you sure?

**Comment by zuntrax** [comment]
The documentation there is fine. How about adding a comment to `libopenage/gamestate/market.cpp` referencing that file?
Also method and class docstrings in doxygen format would be nice.

**Comment by MaanooAk** [comment]
I could add this to the class doc
```c++
/**
 * The global market prices.
 * 
 * Price calculation is documented at doc/reverse_engineering/market.md#prices
 */
class Market {
```
> method and class docstrings in doxygen format

If you mean this:
```c++
 * @param player the player
 * @param res the resource
```
I'm not a big fun of meaningless doc, but If you insist it's ok. Do you insist?

**Comment by zuntrax** [comment]
I'd rather put it in the `cpp` file where it's actually implemented. Or in both the header and `cpp` file.
About the docstrings: I was stupid and looked for them in the `cpp` file and not in the header. The purpose of the parameters is quite obvious, so I think you can skip the `@param` lines. Sorry for the confusion.

**Comment by MaanooAk** [comment]
Source files are not parsed by doxygen (I think), so both seems ok.

File: libopenage/gamestate/market.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+std::shared_ptr<ResourceBundle> Market::getBuyPrices(const Player &player) const {
+	return this->getPrices(player, true);
+}
+
+std::shared_ptr<ResourceBundle> Market::getSellPrices(const Player &player) const {
+	return this->getPrices(player, false);
+}
+
+std::shared_ptr<ResourceBundle> Market::getPrices(const Player &player, const bool is_buy) const {
```
Comment (TheJJ) [inline libopenage/gamestate/market.cpp]: I'm not sure how useful it is to have the resource bundle as `shared_ptr`. Isn't it sufficient as a stack object?

File: libopenage/gamestate/market.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	/**
+	 * The given player buys the given resource with gold.
+	 * Returns true when the transaction is successful.
+	 */
+	bool buy(Player &player, const game_resource res);
+
+	/**
+	 * Get the selling prices for a given player.
+	 */
+	std::shared_ptr<ResourceBundle> getBuyPrices(const Player &player) const;
```
Comment (MaanooAk) [inline libopenage/gamestate/market.h]: I always forget about that (not my fault, I was raised with Java)

Comment (TheJJ) [inline libopenage/gamestate/market.h]: 😀

**Comment by TheJJ** [comment]
I fear you need to rebase because of some fixes we applied in the meantime.

**Pull request merged**